ktbridge/src/jsMain/kotlin/app/cash/quickjs/ktbridge/ktBridgeJs.kt

```diff
@@ -15,28 +15,79 @@
  */
 package app.cash.quickjs.ktbridge
 
+import kotlin.reflect.KClass
 import okio.Buffer
 
+/**
+ * Call this to create a bridge into JavaScript from the JVM.
+ *
+ * The KtBridge Kotlin compiler plugin will rewrite calls to this function to insert a third
+ * parameter to invoke the 3-argument overload.
+ */
 fun <T : Any> createBridgeToJs(service: T, jsAdapter: JsAdapter): BridgeToJs<T> =
   JsBridgeToJs(service, jsAdapter)
 
 private class JsBridgeToJs<T : Any>(val service: T, val jsAdapter: JsAdapter) : BridgeToJs<T> {
   override fun invokeJs(funName: String, arguments: ByteArray): ByteArray {
-    val buffer = Buffer()
-    val eachValueBuffer = Buffer()
-    buffer.write(arguments)
-    val argumentCount = buffer.readInt()
-    val arguments = arrayOfNulls<Any?>(argumentCount)
-    for (i in 0 until argumentCount) {
-      val argumentByteCount = buffer.readInt()
-      eachValueBuffer.write(buffer, argumentByteCount.toLong())
-      val argument = jsAdapter.decode(eachValueBuffer, Any::class)
-      arguments[i] = argument
+    error("unexpected call to invokeJs: is KtBridge plugin configured?")
+  }
+}
+
+/**
+ * Don't invoke this manually; instead call the two-argument form and let the compiler rewrite such
+ * calls to use this.
+ *
+ * This is invoked by compiler-plugin-rewritten code.
+ */
+fun <T : Any> createBridgeToJs(
+  service: T,
+  jsAdapter: JsAdapter,
+  block: (BridgedCall<T>) -> ByteArray
+) : BridgeToJs<T> {
+  return object : BridgeToJs<T> {
+    override fun invokeJs(funName: String, encodedArguments: ByteArray): ByteArray {
+      val bridgedCall = BridgedCall(service, funName, encodedArguments, jsAdapter)
+      return block(bridgedCall)
     }
-    require(buffer.exhausted())
-    @Suppress("UNUSED_VARIABLE") val service = this.service
-    val result = js("""service[funName].apply(service, arguments)""")
-    jsAdapter.encode(result, buffer, Any::class)
+  }
+}
+
+/**
+ * This class models a single call into JavaScript. Each use should call [parameter] once for each
+ * parameter of [funName], then [result] for the function result. This will automatically decode
+ * parameters to the requested type and encode results.
+ *
+ * Call [unexpectedFunction] if an unexpected function is encountered.
+ */
+class BridgedCall<T: Any>(
+  val service: T,
+  val funName: String,
+  encodedArguments: ByteArray,
+  private val jsAdapter: JsAdapter,
+) {
+  private val buffer = Buffer().write(encodedArguments)
+  private val parameterCount = buffer.readInt()
+  private var callCount = 0
+  private val eachParameterBuffer = Buffer()
+
+  inline fun <reified T : Any> parameter(): T = parameter(T::class)
```

Comment:
Need to look at what kotlinx.serialization does, but I suspect we will want to migrate to use `typeOf<T>()` and `KType` rather than `T::class` and `KCalss` so that we can capture generic types.

Reply:
Yes! Nice.

Reply:
(Also I think this might be easier now that everything is generated.)